vault/external_tests/raft/raft_autopilot_test.go

```diff
@@ -412,3 +413,53 @@ func join(t *testing.T, core *vault.TestClusterCore, client *api.Client, cluster
 	time.Sleep(1 * time.Second)
 	cluster.UnsealCore(t, core)
 }
+
+// TestRaft_VotersStayVoters ensures that autopilot doesn't demote a node just
+// because it hasn't been heard from in some time.
+func TestRaft_VotersStayVoters(t *testing.T) {
+	cluster := raftCluster(t, &RaftClusterOpts{
+		DisableFollowerJoins: true,
+		InmemCluster:         true,
+		EnableAutopilot:      true,
+		PhysicalFactoryConfig: map[string]interface{}{
+			"performance_multiplier": "5",
+		},
+		VersionMap: map[int]string{
+			0: version.Version,
+			1: version.Version,
+			2: version.Version,
+		},
+	})
+	defer cluster.Cleanup()
+	testhelpers.WaitForActiveNode(t, cluster)
+
+	client := cluster.Cores[0].Client
+
+	config, err := client.Sys().RaftAutopilotConfiguration()
+	require.NoError(t, err)
+	joinAndStabilizeAndPromote(t, cluster.Cores[1], client, cluster, config, "core-1", 2)
+	joinAndStabilizeAndPromote(t, cluster.Cores[2], client, cluster, config, "core-2", 3)
+
+	errIfNonVotersExist := func() error {
+		t.Helper()
+		resp, err := client.Sys().RaftAutopilotState()
+		if err != nil {
+			t.Fatal(err)
+		}
+		for k, v := range resp.Servers {
+			if v.Status == "non-voter" {
+				return fmt.Errorf("node %q is a non-voter", k)
+			}
+		}
+		return nil
+	}
+	testhelpers.RetryUntil(t, 10*time.Second, errIfNonVotersExist)
+
+	// Core0 is the leader, sealing it will both cause an election - and the
+	// new leader won't have seen any heartbeats initially - and create a "down"
+	// node that won't be sending heartbeats.
+	testhelpers.EnsureCoreSealed(t, cluster.Cores[0])
+	time.Sleep(30 * time.Second)
```

Comment:
Is there a way to replace this with polling?

FWIW I think the cluster will stabilize faster if you change autopilot's update/reconcile intervals down in to the hundred millisecond range instead of leaving them at their defaults. See [here](https://github.com/hashicorp/vault/blob/38abe205ab7bbc5edee022d67a38203d35358f72/vault/external_tests/raft/raft_autopilot_test.go#L208-L209) for an example. Usually 300ms and 100ms work pretty well for the reconcile and update intervals, respectively.

Reply:
Not sure how to use polling here.  We're sleeping to ensure there's time for the new leader to be elected, and for it to decide that the sealed node should be demoted.  If we poll until there's a non-voter, we'll never see it happen given the fix.  If we poll until we see all nodes are voters, the condition will be met immediately, and we'll have to worry about races with reconciliation.

I can change the autopilot intervals, and then see if there's a safe shorter sleep time than 30s, but it feels like a lot of effort and if I get it wrong I risk making the test flaky.  I'd rather try adding t.Parallel to the tests in this file if you want to make it faster - though not in this PR.

Reply:
Sure, that's fine. I don't feel strongly about it.